From edbc93439cb9ee6c5157dcf76a0e2dbe5c75897f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?G=C3=BCnther=20Noack?= <gnoack@google.com>
Date: Mon, 30 Jun 2025 13:00:46 +0200
Subject: [PATCH 09/33] pdftoraster, ghostscript, mupdftopwg: Check strchr()
 result

When processing the %%PDFTOPDFNumCopies and %%PDFTOPDFCollate
comments, these filters should check the result of strchr(),
to check for the case where the colon character is not found,
and to avoid dereferencing the null pointer (or the (NULL+1)-pointer).
---
 cupsfilters/ghostscript.c   | 10 ++++++++++
 cupsfilters/mupdftopwg.c    | 10 ++++++++++
 cupsfilters/pdftoraster.cxx | 10 ++++++++++
 3 files changed, 30 insertions(+)

diff --git a/cupsfilters/ghostscript.c b/cupsfilters/ghostscript.c
index fbcc6c40..30d0e394 100644
--- a/cupsfilters/ghostscript.c
+++ b/cupsfilters/ghostscript.c
@@ -92,6 +92,11 @@ parse_pdf_header_options(FILE *fp,
       char *p;
 
       p = strchr(buf + 19, ':');
+      if (!p)
+      {
+        h->NumCopies = 1;
+        continue;
+      }
       h->NumCopies = atoi(p + 1);
     }
     else if (strncmp(buf, "%%PDFTOPDFCollate", 17) == 0)
@@ -99,6 +104,11 @@ parse_pdf_header_options(FILE *fp,
       char *p;
 
       p = strchr(buf + 17, ':');
+      if (!p)
+      {
+        h->Collate = CUPS_FALSE;
+        continue;
+      }
       while (*p == ' ' || *p == '\t')
 	p ++;
       if (strncasecmp(p, "true", 4) == 0)
diff --git a/cupsfilters/mupdftopwg.c b/cupsfilters/mupdftopwg.c
index 0391a7cf..dcde56ed 100644
--- a/cupsfilters/mupdftopwg.c
+++ b/cupsfilters/mupdftopwg.c
@@ -85,6 +85,11 @@ parse_pdf_header_options(FILE *fp,
       char *p;
 
       p = strchr(buf + 19, ':');
+      if (!p)
+      {
+	h->NumCopies = 1;
+	continue;
+      }
       h->NumCopies = atoi(p + 1);
     }
     else if (strncmp(buf, "%%PDFTOPDFCollate", 17) == 0)
@@ -92,6 +97,11 @@ parse_pdf_header_options(FILE *fp,
       char *p;
 
       p = strchr(buf + 17, ':');
+      if (!p)
+      {
+        h->Collate = CUPS_FALSE;
+        continue;
+      }
       while (*p == ' ' || *p == '\t')
 	p ++;
       if (strncasecmp(p, "true", 4) == 0)
diff --git a/cupsfilters/pdftoraster.cxx b/cupsfilters/pdftoraster.cxx
index 40e16e36..1d14cd59 100644
--- a/cupsfilters/pdftoraster.cxx
+++ b/cupsfilters/pdftoraster.cxx
@@ -429,6 +429,11 @@ parse_pdftopdf_comment(FILE *fp,
       char *p;
 
       p = strchr(buf + 19, ':');
+      if (!p)
+      {
+	(*deviceCopies) = 1;
+	continue;
+      }
       (*deviceCopies) = atoi(p + 1);
     }
     else if (strncmp(buf, "%%PDFTOPDFCollate", 17) == 0)
@@ -436,6 +441,11 @@ parse_pdftopdf_comment(FILE *fp,
       char *p;
 
       p = strchr(buf + 17, ':');
+      if (!p)
+      {
+	*deviceCollate = false;
+	continue;
+      }
       while (*p == ' ' || *p == '\t') p ++;
       if (strncasecmp(p, "true", 4) == 0)
 	*deviceCollate = true;
-- 
2.53.0

